File: kernels/Julia/Compositron/examples/_bench_cdbs.jl
using Compositron.Core
using Compositron.Utils
using Compositron.CDBS

using Statistics


function work()
    import_bench = @timed Measurement(
        "../../../../testdata/depth-profile_Copper_0000.n42"
    )
    m = import_bench.value
    import_time = import_bench.time * 1e6

    c = m.cdbs["OAA x OAB"]

    corr_bench = @timed correct_ecal!(c, EcalCorrOrder_First)
    corr_time = corr_bench.time * 1e6

    param = LineshapeParamDefinition(
        "S",
        [DiagonalArea(KeV(2), KeV(1))],
        [DiagonalArea(KeV(10), KeV(1))],
        false
    )

    s_calc_bench = @timed calc_lineshape_param!(c, param)
    s_calc_time = s_calc_bench.time * 1e6


    proj_diag_bench = @timed project_diagonal(
        c, LinearProjectionBins(KeV(0.1)), KeV(2), KeV(100), false
    )
    proj_diag_time = proj_diag_bench.time * 1e6

    proj_axes_bench = @timed project_axes(c, Axis_FirstDet)
    proj_axes_time = proj_axes_bench.time * 1e6

    (import_time, corr_time, s_calc_time, proj_diag_time, proj_axes_time)
end


function print_mean_and_std(name::String, data::AbstractVector)
    m = mean(data)
    s = std(data)

    println("$name: $m µs +/- $s µs")
end


n_iter = 10000

imports = zeros(Float64, n_iter)
corr = zeros(Float64, n_iter)
s_calc = zeros(Float64, n_iter)
proj_diag = zeros(Float64, n_iter)
proj_axes = zeros(Float64, n_iter)

work()

for i in 1:n_iter
    (import_time, corr_time, s_calc_time, proj_diag_time, proj_axes_time) = work()

    imports[i] = import_time
    corr[i] = corr_time
    s_calc[i] = s_calc_time
    proj_diag[i] = proj_diag_time
    proj_axes[i] = proj_axes_time
end

print_mean_and_std("Import", imports)
print_mean_and_std("Ecal Corr", corr)
print_mean_and_std("S Param", s_calc)
print_mean_and_std("Proj Diag", proj_diag)
print_mean_and_std("Proj Axes", proj_axes)

File: kernels/C++/compositron/examples/_bench_cdbs.cpp
#include "compositron/cdbs/cdbspectrum.hpp"
#include "compositron/core/measurement.hpp"
#include "compositron/core/utils.hpp"
#include <chrono>
#include <cmath>
#include <numeric>
#include <vector>


void print_mean_and_std(std::string name, std::vector<double>& data) {
    const double mean =
        std::accumulate(data.begin(), data.end(), 0.0) / data.size();

    double sum_sq = 0.0;
    for (double x : data) {
        const double diff = x - mean;
        sum_sq += diff * diff;
    }

    const double stddev = std::sqrt(sum_sq / data.size());

    std::cout << name << ": " << mean << " ns +/- " << stddev << " ns\n";
}


int main() {
    using compositron::core::utils::Unit;
    using compositron::core::utils::KEV;

    size_t n_iter = 10000;

    std::vector<double> imports(n_iter);
    std::vector<double> corr(n_iter);
    std::vector<double> s_calc(n_iter);
    std::vector<double> proj_diag(n_iter);
    std::vector<double> proj_axes(n_iter);

    for (size_t i = 0; i < n_iter; ++i) {
        auto start = std::chrono::steady_clock::now();
        compositron::core::Measurement m(
            "../../../../testdata/depth-profile_Copper_0000.n42"
        );
        auto stop = std::chrono::steady_clock::now();
        imports[i] = std::chrono::duration<double, std::micro>(stop - start).count();

        auto& c = m.cdbs.at("OAA x OAB");

        start = std::chrono::steady_clock::now();
        c.correct_ecal(compositron::core::utils::EcalCorrectionOrder::FIRST);
        stop = std::chrono::steady_clock::now();
        corr[i] = std::chrono::duration<double, std::micro>(stop - start).count();

        compositron::cdbs::LineshapeParamDefinition ls_param {
            "S",
            {compositron::cdbs::DiagonalArea(Unit(2, KEV), Unit(1, KEV))},
            {compositron::cdbs::DiagonalArea(Unit(10, KEV), Unit(1, KEV))},
            false
        };

        start = std::chrono::steady_clock::now();
        c.calc_lineshape_param(ls_param);
        stop = std::chrono::steady_clock::now();
        s_calc[i] = std::chrono::duration<double, std::micro>(stop - start).count();

        start = std::chrono::steady_clock::now();
        auto p = c.project_diagonal(
            Unit(0.1, KEV), Unit(2, KEV), Unit(100, KEV), false
        );
        stop = std::chrono::steady_clock::now();
        proj_diag[i] = std::chrono::duration<double, std::micro>(stop - start).count();

        start = std::chrono::steady_clock::now();
        auto x = c.project_axes(compositron::cdbs::Axis::FIRST_DET);
        stop = std::chrono::steady_clock::now();
        proj_axes[i] = std::chrono::duration<double, std::micro>(stop - start).count();
    }

    print_mean_and_std("Import", imports);
    print_mean_and_std("Ecal Corr", corr);
    print_mean_and_std("S Param", s_calc);
    print_mean_and_std("Proj Diag", proj_diag);
    print_mean_and_std("Proj Axes", proj_axes);

    return 0;
}
